askama/src/filters/mod.rs

```diff
@@ -259,25 +262,43 @@ where
 }
 
 /// Joins iterable into a string separated by provided argument
-pub fn join<T, I, S>(input: I, separator: S) -> Result<String>
+#[inline]
+pub fn join<I, S>(input: I, separator: S) -> Result<JoinFilter<I, S>, Infallible>
 where
-    T: fmt::Display,
-    I: Iterator<Item = T>,
-    S: AsRef<str>,
+    I: IntoIterator,
+    I::Item: fmt::Display,
+    S: fmt::Display,
 {
-    let separator: &str = separator.as_ref();
+    Ok(JoinFilter(RefCell::new(Some((input, separator)))))
+}
 
-    let mut rv = String::new();
+/// Result of the filter [`join()`].
+///
+/// ## Note
+///
+/// This struct implements [`fmt::Display`], but only produces a string once.
+/// Any subsequent call to `.to_string()` will result in an empty string, because the iterator is
+/// already consumed.
+#[derive(Debug)]
+pub struct JoinFilter<I, S>(RefCell<Option<(I, S)>>);
```

Comment:
It does not actually need `RefCell`. It only needs RefCell in order to implement `Debug`, but I guess the type does not need to be `Debug`. Changed to `Cell`.

It needs to be a `Cell`, though, because we consume the iterator, and I don't want to force it to be `Clone`.

Reply:
Can you add a comment explaining why it needs to be in a `Cell`?